utf8read.c

```diff
@@ -0,0 +1,77 @@
+#include "hassmtputf8.h"
+
+#ifdef SMTPUTF8
+#include "utf8read.h"
+#include "stralloc.h"
+#include "case.h"
+#include "substdio.h"
+#include "subfd.h"
+
+static stralloc receivedline = { 0 };
+stralloc        header = { 0 };
+extern void temp_nomem();
+extern void temp_read();
+
+int
+containsutf8(unsigned char *p, int l)
+{
+	int             i = 0;
+
+	while (i < l)
+		if (p[i++] > 127)
+			return 1;
+	return 0;
+}
+
+int
+utf8read()
+{
+  int             r, i, received = 0;
+  char            ch;
+
+  if (flagutf8)
+    return 1;
+  /*- initialize headers so that we don't create if called more than once */
+  header.len = receivedline.len = 0;
+  for (;;) {
+    r = substdio_get(subfdin, &ch, 1);
+    if (r == 0) break;
+    if (r == -1) temp_read();
+
+    if (ch == '\n') {
+      if (!stralloc_append(&header, "\r")) temp_nomem(); /* received.c does not add '\r' */
+      if (!stralloc_append(&header, "\n")) temp_nomem();
+      if (case_starts(receivedline.s, "Date:")) return 0;  /* header to quit asap */
+      if (case_starts(receivedline.s, "Received: from")) received++;  /* found Received header */
+      if (received) {
+        for (i = 0; i < receivedline.len; i++)
+          if (*(receivedline.s + i) != ' ' && *(receivedline.s + i) != '\t')
+            break;
+        if (case_starts(receivedline.s + i, "by ")) {
+          for (i += 3; i < receivedline.len; ++i)
+            if (*(receivedline.s + i) == ' ' || *(receivedline.s + i) == '\t')
+              if (case_starts(receivedline.s + i + 1, "with UTF8")) {
+                flagutf8 = 1;return 1;
+              }
+          return 0;
+        }
+        for (i = 0; i < receivedline.len; i++) {
+          if (case_starts(receivedline.s + i, "by ")) {
+            for (i += 3; i < receivedline.len; ++i)
+              if (*(receivedline.s + i) == ' ' || *(receivedline.s + i) == '\t')
+                if (case_starts(receivedline.s + i + 1, "with UTF8")) {
+                  flagutf8 = 1;return 1;
+                }
+            return 0;
+          }
+        }
+      }
+      if (!stralloc_copys(&receivedline, "")) temp_nomem();
+    } else {
+      if (!stralloc_append(&header, &ch)) temp_nomem();
+      if (!stralloc_catb(&receivedline, &ch, 1)) temp_nomem();
+    }
+  }
+  return 0;
+}
+#endif
```

Comment:
The implementation from the patch was using a state machine, rather fun and creative, but plain dumb if/for code like yours feels more straightforward IMHO.

Here it is:

```
+void checkutf8message()
+{
+  int pos;
+  int i;
+  int r;
+  char ch;
+  int state;
+
+  if (containsutf8(sender.s, sender.len)) { utf8message = 1; return; }
+  for (i = 0;i < reciplist.len;++i)
+    if (containsutf8(reciplist.sa[i].s, reciplist.sa[i].len)) {
+      utf8message = 1;
+      return;
+    }
+
+  state = 0;
+  pos = 0;
+  for (;;) {
+    r = substdio_get(&ssin,&ch,1);
+    if (r == 0) break;
+    if (r == -1) temp_read();
+
+    if (!stralloc_append(&firstpart,&ch)) temp_nomem();
+
+    if (ch == '\r')
+      continue;
+    if (ch == '\t')
+      ch = ' ';
+
+    switch (state) {
+    case 6: /* in Received, at LF but before WITH clause */
+      if (ch == ' ') { state = 3; pos = 1; continue; }
+      state = 0;
+      /* FALL THROUGH */
+
+    case 0: /* start of header field */
+      if (ch == '\n') return;
+      state = 1;
+      pos = 0;
+      /* FALL THROUGH */
+
+    case 1: /* partway through "Received:" */
+      if (ch != "RECEIVED:"[pos] && ch != "received:"[pos]) { state = 2; continue; }
+      if (++pos == 9) { state = 3; pos = 0; }
+      continue;
+
+    case 2: /* other header field */
+      if (ch == '\n') state = 0;
+      continue;
+
+    case 3: /* in Received, before WITH clause or partway though " with " */
+      if (ch == '\n') { state = 6; continue; }
+      if (ch != " WITH "[pos] && ch != " with "[pos]) { pos = 0; continue; }
+      if (++pos == 6) { state = 4; pos = 0; }
+      continue;
+
+    case 4: /* in Received, having seen with, before the argument */
+      if (pos == 0 && (ch == ' ' || ch == '\t')) continue;
+      if (ch != "UTF8"[pos] && ch != "utf8"[pos]) { state = 5; continue; }
+      if(++pos == 4) { utf8message = 1; state = 5; continue; }
+      continue;
+
+    case 5: /* after the RECEIVED WITH argument */
+      /* blast() assumes that it copies whole lines */
+      if (ch == '\n') return;
+      state = 1;
+      pos = 0;
+      continue;
+    }
+  }
+}
```

Cannot most of it be summarized by `strncasestr(fieldstart, fieldlen, " with UTF8")` ? Case-insensitive searching  for the string " with utf8" within the "Received:" header.